src/ensembl/src/content/app/browser/zmenu/ZmenuController.tsx

```diff
@@ -29,74 +29,67 @@ const ZmenuController = (props: Props) => {
   const [zmenus, setZmenus] = useState<StateZmenu>({});
 
   useEffect(() => {
-    const browserElement = props.browserRef.current as HTMLDivElement;
-    const eventHandler = (event: Event) =>
-      handleBpaneEvent(event as ZmenuIncomingEvent);
+    const subscriptionHandler = (payload: ZmenuIncomingPayload) =>
+      handleBpaneEvent(payload);
 
-    browserElement.addEventListener('bpane-zmenu', eventHandler);
+    const subscription = browserMessagingService.subscribe(
+      'bpane-zmenu',
+      subscriptionHandler
+    );
 
-    return () =>
-      browserElement.removeEventListener('bpane-zmenu', eventHandler);
+    return () => subscription.unsubscribe();
   }, []);
 
-  const handleBpaneEvent = (event: ZmenuIncomingEvent) => {
-    if (event.detail.action === ZmenuAction.CREATE) {
-      handleZmenuCreate(event as ZmenuCreateEvent);
-    } else if (event.detail.action === ZmenuAction.DESTROY) {
-      handleZmenuDestroy(event as ZmenuDestroyEvent);
-    } else if (event.detail.action === ZmenuAction.REPOSITION) {
-      handleZmenuReposition(event as ZmenuRepositionEvent);
+  const handleBpaneEvent = (payload: ZmenuIncomingPayload) => {
+    if (payload.action === ZmenuAction.CREATE) {
+      handleZmenuCreate(payload);
+    } else if (payload.action === ZmenuAction.DESTROY) {
+      handleZmenuDestroy(payload);
+    } else if (payload.action === ZmenuAction.REPOSITION) {
+      handleZmenuReposition(payload);
     }
   };
 
-  const handleZmenuCreate = (event: ZmenuCreateEvent) => {
-    const zmenuData = event.detail;
+  const handleZmenuCreate = (payload: ZmenuCreatePayload) => {
     const newZmenu = {
-      id: zmenuData.id,
-      anchor_coordinates: zmenuData.anchor_coordinates,
-      content: zmenuData.content
+      id: payload.id,
+      anchor_coordinates: payload.anchor_coordinates,
+      content: payload.content
     };
     setZmenus({
       ...zmenus,
-      [zmenuData.id]: newZmenu
+      [payload.id]: newZmenu
     });
   };
 
-  const handleZmenuDestroy = (event: ZmenuDestroyEvent) => {
-    setZmenus(pickBy(zmenus, (value, key) => key !== event.detail.id));
+  const handleZmenuDestroy = (payload: ZmenuDestroyPayload) => {
+    setZmenus(pickBy(zmenus, (value, key) => key !== payload.id));
```

Comment:
Would it be better clone `zmenus`, delete zmenus[payload.id] and call setZmenus?
By this way we don't have to iterate over it.

Reply:
Cloning implies iteration though.

Which cloning method did you have in mind? `Object.assign({}, zmenus)`?

Iteration is a normal way of doing things within the functional programming paradigm. It shouldn't be a concern unless the structure you iterate over gets very large. In this particular case the number of entries in the `zmenu` state is in single digits. Most likely, `1`.

Reply:
I was thinking of the spread operator but I didn't realise that it implies iteration.
Let's leave this as it is.